components/renku_data_services/users/db.py

```diff
@@ -99,27 +108,56 @@ async def get_or_create_user(self, requested_by: APIUser, id: str) -> UserInfo |
             return user
 
     @only_authenticated
-    async def get_users(self, requested_by: APIUser, email: str | None = None) -> list[UserInfo]:
+    async def get_users(self, requested_by: APIUser, email: str | None = None) -> list[UserWithNamespace]:
         """Get users from the database."""
         if not email and not requested_by.is_admin:
             raise errors.Unauthorized(message="Non-admin users cannot list all users.")
         users = await self._get_users(email)
 
-        is_api_user_missing = not any([requested_by.id == user.id for user in users])
+        is_api_user_missing = not any([requested_by.id == user.user.id for user in users])
 
         if not email and is_api_user_missing:
             api_user_info = await self._add_api_user(requested_by)
             users.append(api_user_info)
         return users
 
-    async def _get_users(self, email: str | None = None) -> list[UserInfo]:
+    async def _get_users(self, email: str | None = None) -> list[UserWithNamespace]:
         async with self.session_maker() as session:
             stmt = select(UserORM)
             if email:
                 stmt = stmt.where(UserORM.email == email)
-            res = await session.execute(stmt)
-            orms = res.scalars().all()
-            return [orm.dump() for orm in orms]
+            stmt = stmt.options(selectinload(UserORM.namespace))
+            result = await session.scalars(stmt)
+            users = result.all()
+
+            for user in users:
+                if user.namespace is None:
+                    raise errors.ProgrammingError(message=f"Cannot find a user namespace for user {id}.")
+
+            return [user.namespace.dump_user() for user in users if user.namespace is not None]
+
+    @only_authenticated
+    async def get_kc_users(self, requested_by: APIUser, email: str | None = None) -> list[UserInfo]:
+        """Get users from the database (Keycloak data only)."""
+        if not email and not requested_by.is_admin:
+            raise errors.Unauthorized(message="Non-admin users cannot list all users.")
+        users = await self._get_kc_users(email)
+
+        is_api_user_missing = not any([requested_by.id == user.id for user in users])
+
+        if not email and is_api_user_missing:
+            api_user_info = await self._add_api_user(requested_by)
+            users.append(api_user_info.user)
+        return users
+
+    async def _get_kc_users(self, email: str | None = None) -> list[UserInfo]:
+        async with self.session_maker() as session:
+            stmt = select(UserORM)
+            if email:
+                stmt = stmt.where(UserORM.email == email)
+            result = await session.scalars(stmt)
+            users = result.all()
+            return [user.dump() for user in users]
```

Comment:
suggestion: These are only used tests. So I think we can eliminate them and use the new methods that return "UserWithNamespace" in the tests.

Reply:
Done in 09b2d7a64537057c618ce686894d3efc8e09fef4.